app/controllers/application_controller.rb

```diff
@@ -275,9 +281,12 @@ def authenticated_as_user?(user, reason_params = {})
 
   # Return logged in user
   def authenticated_user
-    return unless session[:user_id]
+    return unless session[:user_id] && session[:user_last_login_at]
 
-    @user ||= User.find_by(id: session[:user_id])
+    @user ||= (
+      user = User.find_by(id: session[:user_id])
+      user if user&.last_login_at.to_s == session[:user_last_login_at]
```

Comment:
Spec needed to show what is returned is these do and don't match

Reply:
Comparing string representation of the timestamp, should we compare actual timestamps?

I went for strings here as this quickly dealt with milliseconds differences.

Reply:
Do we want to do the check as part of the ActiveRecord query?

Reply:
> Do we want to do the check as part of the ActiveRecord query?

Something like that seems quite nice:

```ruby
# A
@user ||= User.find_logged_in(session[:user_id], session[:user_last_login_at])

# B
@user ||= User.logged_in(session[:user_last_login_at]).find(session[:user_id])
```

A is better in the context of the controller since it reduces the method chain, but it's definition in `User` could be:

```ruby
def self.find_logged_in(id, last_login_at)
  logged_in(last_login_at).find(id)
end
```

Reply:
Doing time comparison in PG is such a pain. Though I had is sorted a couple of times but with either milliseconds or time zone issues it would break in different environments.

I've gone with doing the comparison in Ruby but have extracted to a `User.find_logged_in` method instead of doing it in the `ApplicationController`.